### ISSUE TITLE
Error running MeshTools::Genertion::build_sphere() !

### ISSUE THREAD

**Issue by ermore** [libMesh/libmesh#2496]
When I run the following code:
```c++
MeshTools::Generation::build_sphere (mesh, 10.0, 2, HEX27, 2, true);
```
I get an error:
```
[0]PETSC ERROR: ------------------------------------------------------------------------
[0]PETSC ERROR: Caught signal number 11 SEGV: Segmentation Violation, probably memory access out of range
[0]PETSC ERROR: Try option -start_in_debugger or -on_error_attach_debugger
[0]PETSC ERROR: or see https://www.mcs.anl.gov/petsc/documentation/faq.html#valgrind
[0]PETSC ERROR: or try http://valgrind.org on GNU/linux and Apple Mac OS X to find memory corruption errors
[0]PETSC ERROR: configure using --with-debugging=yes, recompile, link, and run 
[0]PETSC ERROR: to get more information on the crash.
[0]PETSC ERROR: --------------------- Error Message --------------------------------------------------------------
[0]PETSC ERROR: Signal received
[0]PETSC ERROR: See https://www.mcs.anl.gov/petsc/documentation/faq.html for trouble shooting.
[0]PETSC ERROR: Petsc Release Version 3.12.1, Oct, 22, 2019 
[0]PETSC ERROR: ./example-dbg on a arch-linux2-c-opt named ermore by ermore Sun Apr  5 09:26:57 2020
[0]PETSC ERROR: Configure options -with-debugging=0 --download-hdf5 --download-hypre --download-metis --download-parmetis --download-ptscotch --download-zoltan --download-mumps --download-scalapack --download-suitesparse --download-superlu --download-superlu_dist --download-slepc --download-sundials --download-silo --download-zliba
[0]PETSC ERROR: #1 User provided function() line 0 in  unknown file
--------------------------------------------------------------------------
MPI_ABORT was invoked on rank 0 in communicator MPI_COMM_WORLD
with errorcode 59.

NOTE: invoking MPI_ABORT causes Open MPI to kill all MPI processes.
You may or may not see output from other processes, depending on
exactly when Open MPI kills them.
```

**Comment by roystgnr** [comment]
We don't seem to have any test coverage for that function.  Let me add some real quick and see if I can replicate this.

In the meantime - are you running in dbg mode?  Regardless of regressions I'd hope we at *least* hit an assert before getting a segmentation fault.

**Comment by roystgnr** [comment]
Not seeing any regressions with ReplicatedMesh yet, but it's definitely broken with DistributedMesh.  What's confusing is that I don't remember actually ever *using* build_sphere() with DistributedMesh, but there are some commented DistributedMesh-directed updates in the build_sphere code, so I'm not sure whether it was working at one point and regressed or whether the updates were never completed
...
or whether it works in some cases but not others?  It looks like I might have tested only the most superficial cases but not tested it with refinements, which get followed by a MeshModification::flatten() call, which doesn't seem to have been updated to support an already-distributed mesh.  I'll see how hard that will be to fix.

**Comment by roystgnr** [comment]
No, flatten() was updated.  Still not sure where the bug is, then.

**Comment by roystgnr** [comment]
I believe this should be fixed as of #2503
